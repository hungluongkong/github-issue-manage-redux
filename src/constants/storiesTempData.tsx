/**
 * Template response json string from github.
 * This is a list of issues.
 * URL: https://api.github.com/repos/hungluong-asnet/demo-api/issues
 */
export const issueJson: string = `
{
  "url": "https://api.github.com/repos/hungluong-asnet/demo-api/issues/1",
  "repository_url": "https://api.github.com/repos/hungluong-asnet/demo-api",
  "labels_url": "https://api.github.com/repos/hungluong-asnet/demo-api/issues/1/labels{/name}",
  "comments_url": "https://api.github.com/repos/hungluong-asnet/demo-api/issues/1/comments",
  "events_url": "https://api.github.com/repos/hungluong-asnet/demo-api/issues/1/events",
  "html_url": "https://github.com/hungluong-asnet/demo-api/issues/1",
  "id": 653740906,
  "node_id": "MDU6SXNzdWU2NTM3NDA5MDY=",
  "number": 1,
  "title": "isssue no 1",
  "user": {
    "login": "hungluong-asnet",
    "id": 68043715,
    "node_id": "MDQ6VXNlcjY4MDQzNzE1",
    "avatar_url": "https://avatars1.githubusercontent.com/u/68043715?v=4",
    "gravatar_id": "",
    "url": "https://api.github.com/users/hungluong-asnet",
    "html_url": "https://github.com/hungluong-asnet",
    "followers_url": "https://api.github.com/users/hungluong-asnet/followers",
    "following_url": "https://api.github.com/users/hungluong-asnet/following{/other_user}",
    "gists_url": "https://api.github.com/users/hungluong-asnet/gists{/gist_id}",
    "starred_url": "https://api.github.com/users/hungluong-asnet/starred{/owner}{/repo}",
    "subscriptions_url": "https://api.github.com/users/hungluong-asnet/subscriptions",
    "organizations_url": "https://api.github.com/users/hungluong-asnet/orgs",
    "repos_url": "https://api.github.com/users/hungluong-asnet/repos",
    "events_url": "https://api.github.com/users/hungluong-asnet/events{/privacy}",
    "received_events_url": "https://api.github.com/users/hungluong-asnet/received_events",
    "type": "User",
    "site_admin": false
  },
  "labels": [

  ],
  "state": "open",
  "locked": false,
  "assignee": null,
  "assignees": [

  ],
  "milestone": null,
  "comments": 0,
  "created_at": "2020-07-09T03:48:36Z",
  "updated_at": "2020-07-09T03:48:36Z",
  "closed_at": null,
  "author_association": "OWNER",
  "active_lock_reason": null,
  "body": "Nothing inside issue"
}
`;

export const issueListJson: string = `
  [
    {
      "url": "https://api.github.com/repos/hungluong-asnet/demo-api/issues/2",
      "repository_url": "https://api.github.com/repos/hungluong-asnet/demo-api",
      "labels_url": "https://api.github.com/repos/hungluong-asnet/demo-api/issues/2/labels{/name}",
      "comments_url": "https://api.github.com/repos/hungluong-asnet/demo-api/issues/2/comments",
      "events_url": "https://api.github.com/repos/hungluong-asnet/demo-api/issues/2/events",
      "html_url": "https://github.com/hungluong-asnet/demo-api/issues/2",
      "id": 654622112,
      "node_id": "MDU6SXNzdWU2NTQ2MjIxMTI=",
      "number": 2,
      "title": "#practice #make #api",
      "user": {
        "login": "hungluong-asnet",
        "id": 68043715,
        "node_id": "MDQ6VXNlcjY4MDQzNzE1",
        "avatar_url": "https://avatars1.githubusercontent.com/u/68043715?v=4",
        "gravatar_id": "",
        "url": "https://api.github.com/users/hungluong-asnet",
        "html_url": "https://github.com/hungluong-asnet",
        "followers_url": "https://api.github.com/users/hungluong-asnet/followers",
        "following_url": "https://api.github.com/users/hungluong-asnet/following{/other_user}",
        "gists_url": "https://api.github.com/users/hungluong-asnet/gists{/gist_id}",
        "starred_url": "https://api.github.com/users/hungluong-asnet/starred{/owner}{/repo}",
        "subscriptions_url": "https://api.github.com/users/hungluong-asnet/subscriptions",
        "organizations_url": "https://api.github.com/users/hungluong-asnet/orgs",
        "repos_url": "https://api.github.com/users/hungluong-asnet/repos",
        "events_url": "https://api.github.com/users/hungluong-asnet/events{/privacy}",
        "received_events_url": "https://api.github.com/users/hungluong-asnet/received_events",
        "type": "User",
        "site_admin": false
      },
      "labels": [

      ],
      "state": "open",
      "locked": false,
      "assignee": null,
      "assignees": [

      ],
      "milestone": null,
      "comments": 0,
      "created_at": "2020-07-10T08:47:16Z",
      "updated_at": "2020-07-10T08:47:16Z",
      "closed_at": null,
      "author_association": "OWNER",
      "active_lock_reason": null,
      "body": "**Don't know why this issue is here. and why we need to create it?**"
    },
    ${issueJson}
  ]`;
