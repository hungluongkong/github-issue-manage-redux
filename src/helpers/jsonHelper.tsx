/* eslint-disable camelcase */
import { Issue } from '../types';

export interface RawIssue {
  id: number;
  number: number;
  title: string;
  body: string;
  locked: boolean;
  url: string;
  created_at: string;
  updated_at: string;
}

export const RawToIssue = (raws: RawIssue[]): Issue[] =>
  raws
    .map((raw: RawIssue) => {
      const createdAt = new Date(raw.created_at);
      const updatedAt = new Date(raw.updated_at);

      // Destructor - remove unused properties
      const { id, number, url, title, body, locked } = raw;

      // Create Item
      return {
        id,
        number,
        title,
        body,
        locked,
        url,
        createdAt,
        updatedAt,
      } as Issue;
    })
    .sort((issue: Issue, nextIssue: Issue) => issue.number - nextIssue.number);

/**
 * Convert to list of issue from json
 * @param json response from github api
 */
export const JsonToIssueList = (json: string): Issue[] => {
  const raws: RawIssue[] = JSON.parse(json);

  return RawToIssue(raws);
};
