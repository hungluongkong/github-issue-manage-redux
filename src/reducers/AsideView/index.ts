import Action from '../../constants/actions';
import { Issue, AsideState, AsideActionPayload } from '../../types';

export const defaultState: AsideState = {
  isShowForm: false,
  isShowIssue: false,
  issue: {} as Issue,
  error: '',
};

const asideReducer = (
  state: AsideState = defaultState,
  action: AsideActionPayload,
): AsideState => {
  switch (action.type) {
    case Action.TOGGLE_FORM:
      return {
        ...state,
        isShowIssue: false,
        isShowForm: !state.isShowForm,
        issue: action.issue || ({} as Issue),
      };

    case Action.TOGGLE_ISSUE:
      return {
        ...state,
        isShowForm: false,
        isShowIssue: action.issue?.id !== state.issue.id || !state.isShowIssue,
        issue: action.issue || ({} as Issue),
      };

    case Action.TOGGLE_ERROR:
      return {
        ...state,
        error: action.error || '',
      };
    default:
      return state;
  }
};

export default asideReducer;
