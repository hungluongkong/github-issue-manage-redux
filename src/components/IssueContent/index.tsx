/* -------------------------------------------- */
/*     Display all issue info for one Issue     */
/* -------------------------------------------- */

import React from 'react';
import styled from 'styled-components';
// import { Marked } from '@ts-stack/markdown';

import Button from '../Button';

import * as localize from '../../constants/localize';
import * as color from '../../theme/color';
import { ButtonType, IssueContentProps } from '../../types';

const ContentContainer = styled.div`
  padding: 5px;
  border-radius: 3px;
  background-color: ${color.ISSUE_BG_CONTENT};
`;

const Title = styled.div`
  padding: 10px 0 0 10px;
  font-weight: 700;
  display: block;
  box-shadow: 2px 1px 2px 1px ${color.SHADOW_BORDER};
`;

// Body of content. Contain edit item.
const IssueContentBody = styled.div`
  padding: 10px 0 0 10px;
`;

// Footer of content. Contain edit item.
const IssueContentFooter = styled.div`
  display: flex;
  width: 100%;
  justify-content: flex-end;
`;

const IssueContent = ({
  issue,
  onEditPressed,
}: IssueContentProps) => (
  issue.id ?
    <ContentContainer>
      <Title>{issue.title}</Title>
      <IssueContentBody dangerouslySetInnerHTML={{__html: (issue.body)}} />
      <IssueContentFooter>
        <Button
          type={ButtonType.Default}
          onClick={onEditPressed}
          value={localize.en.BTN_ISSUE_EDIT}
        />
      </IssueContentFooter>
    </ContentContainer>
    :
    // Render nothing
    <React.Fragment />
);

export default IssueContent;
