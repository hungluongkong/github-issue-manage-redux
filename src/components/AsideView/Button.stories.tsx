import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';

import { PreviewArea, Code, DetailTable } from '../commonStyled';
import AsideView from '.';
import { JsonToIssueList } from '../../helpers/jsonHelper';
import { issueListJson } from '../../constants/storiesTempData';

// Detail of props
const propsDetail: JSX.Element = DetailTable({
  data: [
    { name: 'isShowForm', type: 'boolean' },
    { name: 'isShowIssue', type: 'boolean' },
    { name: 'issue', type: 'Issue', description: 'keep empty for empty form' },
    { name: 'toggleForm', type: 'function' },
    { name: 'toggleIssue', type: 'function' },
    { name: 'updateIssue', type: 'function' },
  ],
});

const [issue] = JsonToIssueList(issueListJson);

storiesOf('Aside View', module)
  .addDecorator((story) => <div style={{ margin: '0 30px' }}>{story()}</div>)
  .add('With Form', () => (
    <>
      <h3>With Form</h3>
      <PreviewArea>
        <AsideView
          isShowForm
          isShowIssue={false}
          issue={issue}
          toggleForm={action('toggle form')}
          toggleIssue={action('toggle Issue')}
          updateIssue={action('update Issue')}
        />
      </PreviewArea>
      <Code>
        {`
          <AsideView
            isShowForm
            isShowIssue={false}
            issue={issue}
            toggleForm={action('toggle form')}
            toggleIssue={action('toggle Issue')}
            updateIssue={action('update Issue')}
          />
        `}
      </Code>
      {propsDetail}
    </>
  ))
  .add('With Issue', () => (
    <>
      <h3>With Issue</h3>
      <PreviewArea>
        <AsideView
          isShowForm={false}
          isShowIssue
          issue={issue}
          toggleForm={action('toggle form')}
          toggleIssue={action('toggle Issue')}
          updateIssue={action('update Issue')}
        />
      </PreviewArea>
      <Code>
        {`
          <AsideView
            isShowForm={false}
            isShowIssue
            issue={issue}
            toggleForm={action('toggle form')}
            toggleIssue={action('toggle Issue')}
            updateIssue={action('update Issue')}
          />
        `}
      </Code>
      {propsDetail}
    </>
  ));
