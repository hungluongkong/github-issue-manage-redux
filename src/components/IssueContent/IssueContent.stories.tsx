import React from 'react';

import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { JsonToIssueList } from '../../helpers/jsonHelper';
import { issueListJson } from '../../constants/storiesTempData';
import { PreviewArea, Code, DetailTable } from '../commonStyled';
import IssueContent from '.';

// Detail of props
const propsDetail: JSX.Element = DetailTable({
  data: [{ name: 'issue', type: 'interface Issue' }],
});

const [issue] = JsonToIssueList(issueListJson);

storiesOf('Issue Content', module).add(
  'Default',
  (): JSX.Element => (
    <>
      <h3>Default Item</h3>
      <PreviewArea>
        <IssueContent issue={issue} onEditPressed={action('Edit pressed')} />
      </PreviewArea>
      <Code>
        {`
          <IssueContent issue={issue} />
        `}
      </Code>
      {propsDetail}
    </>
  ),
);
