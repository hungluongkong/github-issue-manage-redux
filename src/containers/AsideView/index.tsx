import { connect } from 'react-redux';
import { RootReducerState } from '../../types';
import AsideView from '../../components/AsideView';
import { updateIssue } from '../../actions/issues';
import { toggleIssue, toggleForm } from '../../actions/asideView';

const mapStateToProps = (state: RootReducerState) => ({
  isShowIssue: state.asideView.isShowIssue,
  isShowForm: state.asideView.isShowForm,
  issue: state.asideView.issue,
});

const mapDispatchToProps = {
  updateIssue,
  toggleIssue,
  toggleForm,
};

export default connect(mapStateToProps, mapDispatchToProps)(AsideView);
