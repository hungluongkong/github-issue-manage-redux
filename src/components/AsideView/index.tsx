import React from 'react';
import styled from 'styled-components';
import IssueContent from '../IssueContent';
import IssueForm from '../Form';
import { Issue, AsideViewProps } from '../../types';
import { isEqualAllProps } from '../../helpers/propsHelper';

const IssueViewWrapper = styled.div`
  width: 40%;
`;

const AsideView = ({
  isShowIssue,
  isShowForm,
  issue,
  updateIssue,
  toggleIssue,
  toggleForm,
}: AsideViewProps) => (
  // Only show if it have show form or issue content
  isShowIssue || isShowForm ?
    <IssueViewWrapper>
      {
      // Show form (Edit)
        isShowIssue ?
          (
            issue && issue.id &&
            <IssueContent
              issue={issue}
              onEditPressed={() => {
                toggleForm(issue);
              }}
            />
          )
          :
          (
            <IssueForm
              issue={issue}
              onSubmit={(issue: Issue) => {
                updateIssue(issue);
                toggleIssue(issue);
              }}
            />
          )
      }
    </IssueViewWrapper>
    :
    // Show nothing
    <React.Fragment />
);

export default React.memo(AsideView, isEqualAllProps);
