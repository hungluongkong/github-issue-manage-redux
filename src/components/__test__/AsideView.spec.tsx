import React from 'react';
import { create } from 'react-test-renderer';
import { shallow } from 'enzyme';
import AsideView from '../AsideView';
import IssueForm from '../Form';
import IssueContent from '../IssueContent';
import { AsideViewProps, Issue } from '../../types';
import { JsonToIssueList } from '../../helpers/jsonHelper';
import { issueListJson } from '../../constants/storiesTempData';

describe('Unit test for AsideView component', () => {
  const [issue]: Issue[] = JsonToIssueList(issueListJson);

  const defaultProps: AsideViewProps = {
    isShowForm: false,
    isShowIssue: false,
    issue,

    // Events should be test by child components.
    toggleForm: jest.fn(),
    toggleIssue: jest.fn(),
    updateIssue: jest.fn(),
  };

  it('snapshot', () => {
    const asideView = create(<AsideView {...{...defaultProps, isShowIssue: true}} />);
    expect(asideView).toMatchSnapshot();
  });

  it('should show nothing', () => {
    const asideView = create(<AsideView {...defaultProps} />).toJSON();
    expect(asideView).toBeNull();
  });

  it('should show form', () => {
    const asideView = shallow(<AsideView {...{...defaultProps, isShowForm: true}} />);
    expect(asideView.find(IssueForm).exists()).toEqual(true);
  });

  it('should show issue content', () => {
    const asideView = shallow(<AsideView {...{...defaultProps, isShowIssue: true}} />);
    expect(asideView.find(IssueContent).exists()).toEqual(true);
  });
});
