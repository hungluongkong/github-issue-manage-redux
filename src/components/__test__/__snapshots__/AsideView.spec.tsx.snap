// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit test for AsideView component snapshot 1`] = `
<div
  className="sc-fzpans eyWCTA"
>
  <div
    className="sc-AxirZ bNsBBv"
  >
    <div
      className="sc-AxiKw jBSfnA"
    >
      isssue no 1
    </div>
    <div
      className="sc-AxhCb hGUfeO"
      dangerouslySetInnerHTML={
        Object {
          "__html": "Nothing inside issue",
        }
      }
    />
    <div
      className="sc-AxhUy gPNCHG"
    >
      <button
        className="sc-AxjAm gZMYpY"
        onClick={[Function]}
        size={1}
      >
        Edit
      </button>
    </div>
  </div>
</div>
`;
